x-pack/plugins/enterprise_search/public/applications/app_search/components/role_mappings/role_mappings_logic.ts

```diff
@@ -0,0 +1,337 @@
+    elasticsearchRoles: [
+      [],
+      {
+        setRoleMappingData: (_, { elasticsearchRoles }) => elasticsearchRoles,
+      },
+    ],
+    roleMapping: [
+      null,
+      {
+        setRoleMappingData: (_, { roleMapping }) => roleMapping || null,
+        resetState: () => null,
+      },
+    ],
+    roleType: [
+      'owner',
+      {
+        setRoleMappingData: (_, { roleMapping }) =>
+          roleMapping ? (roleMapping.roleType as RoleTypes) : 'owner',
+        handleRoleChange: (_, { roleType }) => roleType,
+      },
+    ],
+    accessAllEngines: [
+      true,
+      {
+        setRoleMappingData: (_, { roleMapping }) =>
+          roleMapping ? roleMapping.accessAllEngines : true,
+        handleRoleChange: (_, { roleType }) => !roleHasScopedEngines(roleType),
+        handleAccessAllEnginesChange: (accessAllEngines) => !accessAllEngines,
+      },
+    ],
+    attributeValue: [
+      '',
+      {
+        setRoleMappingData: (_, { roleMapping }) =>
+          roleMapping ? getFirstAttributeValue(roleMapping) : '',
+        handleAttributeSelectorChange: (_, { value, firstElasticsearchRole }) =>
+          value === 'role' ? firstElasticsearchRole : '',
+        handleAttributeValueChange: (_, { value }) => value,
+        resetState: () => '',
+      },
+    ],
+    attributeName: [
+      'username',
+      {
+        setRoleMappingData: (_, { roleMapping }) =>
+          roleMapping ? getFirstAttributeName(roleMapping) : 'username',
+        handleAttributeSelectorChange: (_, { value }) => value,
+        resetState: () => 'username',
+      },
+    ],
+    selectedEngines: [
+      new Set(),
+      {
+        setRoleMappingData: (_, { roleMapping }) =>
+          roleMapping ? new Set(roleMapping.engines.map((engine) => engine.name)) : new Set(),
+        handleAccessAllEnginesChange: () => new Set(),
+        handleEngineSelectionChange: (engines, { engineName, selected }) => {
+          const newSelectedEngineNames = new Set(engines as Set<string>);
+          if (selected) {
+            newSelectedEngineNames.add(engineName);
+          } else {
+            newSelectedEngineNames.delete(engineName);
+          }
+
+          return newSelectedEngineNames;
+        },
+      },
+    ],
+    availableAuthProviders: [
+      [],
+      {
+        setRoleMappingData: (_, { authProviders }) => authProviders,
+      },
+    ],
+    selectedAuthProviders: [
+      [ANY_AUTH_PROVIDER],
+      {
+        handleAuthProviderChange: (previous, { value }) => {
+          const previouslyContainedAny = previous.includes(ANY_AUTH_PROVIDER);
+          const newSelectionsContainAny = value.includes(ANY_AUTH_PROVIDER);
+
+          if (value.length < 1) return [ANY_AUTH_PROVIDER];
+          if (value.length === 1) return value;
+          if (!newSelectionsContainAny) return value;
+          if (previouslyContainedAny) return value.filter((v) => v !== ANY_AUTH_PROVIDER);
+          return [ANY_AUTH_PROVIDER];
+        },
```

Comment:
This was a bit of a doozy to parse out lol, I'm still not 100% I've fully grasped it.

Just for my own edification, is there a realistic scenario where this falls through to line 236 / `return [ANY_AUTH_PROVIDER];`? If not, I'm wondering if we should delete line 232 and have that be the fallthrough 🤷

Reply:
I also have not totally grasped it, but it seems to work fine. That's a good question. Because we are going to rewrite this from scratch post 8.0 I vote to leave it as-is